url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/RuleCollection.java

```diff
@@ -0,0 +1,862 @@
+	public static double GLOBAL_POSITIVE_MATCH_THRESHOLD = .9;
+    /**The threshold that must be passed to meet the positive match criterion.*/
+    public double POSITIVE_MATCH_THRESHOLD = GLOBAL_POSITIVE_MATCH_THRESHOLD;
+	
+	/**
+	 * Initializes a rule collection.
+	 */
+	public RuleCollection ()
+	{
+		super();
+	}
+	
+	/**
+	 * Initializes a rule collection with the collection of rules specified.
+	 * @param rules The rules for the collection.
+	 */
+	public RuleCollection (Collection <? extends AbstractRule> rules)
+	{
+		super();
+		for(AbstractRule r : rules)
+			add(r);
+	}
+	
+	/**
+	 * Gets a rule in the collection that matches the specified condition and 
+	 * action. If the collection does not contain a rule with the specified condition and 
+	 * action, this method returns null.
+	 * NOTE: If the specified condition of the rule to get is null (such as in the case of a fixed rule without a condition), 
+	 * this method will return the condition-less rule for the specified action.
+	 * @param cond The condition of the rule to get.
+	 * @param act The action of the rules to get.
+	 * @return A rules with the specified condition and action.
+	 */
+	public AbstractRule get (GeneralizedConditionChunk cond, AbstractAction act)
+	{
+		HashMap<Object, HashMap<Object,LinkedList <AbstractRule>>> rc = get(act.getID());
+		if (rc == null)
+			return null;
+		if(cond == null)
+		{
+			//Only 1 condition-less rule is allowed for each action. This is only the case for a rule 
+			//that does not have a condition.
+			LinkedList <AbstractRule> candidates = rc.get(null).get(null);
+			if(candidates == null)
+				return null;
+			return candidates.iterator().next();
+		}
+		else
+		{
+			for(Dimension d : cond.values())
+			{
+				HashMap<Object,LinkedList <AbstractRule>> rrc = rc.get(d.getID());
+				if(rrc == null)
+					return null;
+				for(Value v : d.values())
+				{
+					if(v.isFullyActivated())
+					{
+						LinkedList <AbstractRule> candidates = rrc.get(v.getID());
+						if(candidates == null)
+							return null;
+						for(AbstractRule r : candidates)
+						{
+							if(r.getCondition().equals(cond))
+								return r;
+						}
+						return null;
+					}
+				}
+			}
+		}
+		//It is highly unlikely this return statement will be reached. The specified condition would have to 
+		//have no dimensions OR all of the values within all of the dimensions of the specified condition
+		//would have to be "off", in which case the condition would never match any state and would therefore
+		//not be very useful.
+		return null;
+	}
+	
+	/**
+	 * Gets the rules in the collection that cover the specified rule. If the collection does 
+	 * not contain any rules that cover the specified rule, this method returns null.
+	 * @param R The rule you wish to check for the existence of a covering rule within the collection.
+	 * @return A collection of rules that cover the specified rule.
+	 */
+	public Collection <AbstractRule> getCover (AbstractRule R)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.